src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -299,6 +299,122 @@ static void insertConvSpatialDim(SmallVector<int64_t, 4> *outputDims,
   }
 }
 
+//===----------------------------------------------------------------------===//
+// Support function that infers shape for RNN operations.
+template <typename T>
+static bool RNNShapeInference(T *op) {
+  Value X = op->X();
+  Value W = op->W();
+  Value R = op->R();
+
+  if (!X.getType().isa<RankedTensorType>() ||
+      !W.getType().isa<RankedTensorType>() ||
+      !R.getType().isa<RankedTensorType>())
+    return false;
+
+  auto xTy = X.getType().cast<RankedTensorType>();
+  auto elementType = xTy.getElementType();
+
+  // xShape :: [seq_length, batch_size, input_size]
+  auto xShape = xTy.getShape();
+  // wShape :: [num_directions, 4*hidden_size, input_size]
+  auto wShape = W.getType().cast<RankedTensorType>().getShape();
+  // rShape :: [num_directions, 4*hidden_size, hidden_size]
+  auto rShape = R.getType().cast<RankedTensorType>().getShape();
+
+  if (xShape.size() != 3) {
+    op->emitError("The first input tensor must have rank 3");
+    return false;
+  }
+  if (wShape.size() != 3) {
+    op->emitError("The second input tensor must have rank 3");
+    return false;
+  }
+  if (rShape.size() != 3) {
+    op->emitError("The third input tensor must have rank 3");
+    return false;
+  }
+
+  // Get sequence length, batch size and input size.
+  auto sequenceLength = xShape[0];
+  auto batchSize = xShape[1];
+  auto inputSize = xShape[2];
+
+  // Get hidden size from hidden_size attribute.
+  int64_t hiddenSize = -1;
+  if (op->hidden_size().hasValue()) {
+    hiddenSize = op->hidden_size().getValue().getSExtValue();
+  } else {
+    // Infer hidden_size from wShape and rShape if possible.
+    if (rShape[2] != -1)
+      hiddenSize = rShape[2];
+    else if (rShape[1] != -1)
+      hiddenSize = rShape[1] / 4;
+    else if (wShape[1] != -1)
+      hiddenSize = wShape[1] / 4;
+    // Update hidden_size attribute.
+    if (hiddenSize != -1) {
+      auto builder = mlir::Builder(op->getContext());
+      op->hidden_sizeAttr(builder.getI64IntegerAttr(hiddenSize));
+    }
+  }
+
+  // Get direction.
+  int numDirection;
+  if ((op->direction() == "forward") || (op->direction() == "reverse"))
+    numDirection = 1;
+  else if (op->direction() == "bidirectional")
+    numDirection = 2;
+  else
+    numDirection = -1;
+  if (numDirection == -1) {
+    op->emitError("direction attribute muse be one of the strings: forward, "
+                  "reverse, and bidirectional");
+    return false;
+  }
+
+  // Set result types.
+  unsigned numOfResults = op->getNumResults();
+  if (numOfResults > 0) {
+    // Y :: [seq_length, num_directions, batch_size, hidden_size]
+    Type yTy = op->getResults()[0].getType();
+    if (!yTy.isa<NoneType>()) {
+      SmallVector<int64_t, 4> yDims;
+      yDims.emplace_back(sequenceLength);
```

Comment:
Again, literal assignment might make things easier to read. Here we may go directly to
```
RankedTensorType::get({sequenceLength, numDirection, batchSize, hiddenSize}, elementType);
```